File: services/downloaders/create-dept-metadata.js
import path from 'path';
import * as url from "url";
import {readFileSync, writeFileSync, existsSync, mkdirSync, readdirSync, statSync, stat} from 'fs';
import axios from 'axios';
import * as cheerio from 'cheerio';

import { prepend } from '../utils/url-utils.js'
import { transformDateFormat } from '../utils/date-utils.js'
import { downloadPDF, downloadPDFFile } from "../utils/download-pdf.js";

const directoryUrl = new URL(".", import.meta.url);
const __dirname = url.fileURLToPath(directoryUrl);
const parentDir = path.dirname(__dirname);
console.log(parentDir)

function sleep(ms) {
  return new Promise(resolve => setTimeout(resolve, ms));
}

const placeInfo = {
  name: "jacksonville-fl",
  display: "Jacksonville, FL"
}

const placeMetaDataFolder = path.join(parentDir, 'downloaders', 'metadata', placeInfo.name); 
const placeDataFolder = path.join(parentDir, '../data', placeInfo.name); 
if(!existsSync(placeDataFolder)) {
  mkdirSync(placeDataFolder)
}
 
const DEPARTMENT_LINKS = {

}

const processMeetings = async (meetings, folderPath) => {
  for(const meeting of meetings) {
    const minutes = meeting["Minutes"]
    if(typeof minutes === "string" || !minutes.link) continue
    console.log(minutes.link)
    const department = meeting["Department"]
    if(!department) {
      continue
    }
    const deptName = department.value || department
    const deptDataFolder = path.join(folderPath, deptName);
    if(!existsSync(deptDataFolder)) {
      mkdirSync(deptDataFolder)
    }
  if(department.link) {
    DEPARTMENT_LINKS[deptName] = department.link 
  }        
  const dateStr = transformDateFormat( meeting["Date"])
  const timeStr = meeting["Time"].split(' ')[0].replace(':', '-')
    const minutesFilePath = path.join(deptDataFolder, `${dateStr}_${timeStr}_minutes.pdf`)
    await downloadPDFFile(minutes.link, minutesFilePath)
  }
}

const processPlaceMetadata = async (placeMetaDataFolder, placeDataFolder) => {
  const metadataFilePath = path.join(placeMetaDataFolder, 'download-metadata.json');
  if (existsSync(metadataFilePath)) {
    // Read the metadata.json file
    const data = readFileSync(metadataFilePath, 'utf8');
    // Parse the JSON data
    const metadata = await JSON.parse(data);
    console.log(`Processing ${metadata.place}`)

    for(const department of Object.keys(metadata)) {
      if (department === 'place') {
        continue
      }
      if(metadata[department].meetings) {
        processMeetings(metadata[department].meetings, placeDataFolder)
      }
    }
    await sleep(1000)
  } else {
    console.warn(`metadata.json not found in ${placeMetaDataFolder}`);
  }  
} 

const process = async () => {
  console.log('creating metadata for each department');
  await processPlaceMetadata(placeMetaDataFolder, placeDataFolder)
}

process()

File: services/downloaders/create-dept-metadata.py
import os
import json
import time
import urllib.parse
from pathlib import Path

import requests

from ..utils.date_utils import transform_date_format
from ..utils.download_pdf import download_pdf_file

# Get the current directory and parent directory
current_dir = Path(__file__).resolve().parent
parent_dir = current_dir.parent

print(parent_dir)

def sleep(ms):
    time.sleep(ms / 1000)

place_info = {
    "name": "jacksonville-fl",
    "display": "Jacksonville, FL"
}

place_metadata_folder = parent_dir / 'crawlers' / 'entity' / place_info['name']
place_data_folder = parent_dir.parent / 'data' / place_info['name']
place_data_folder.mkdir(parents=True, exist_ok=True)

DEPARTMENT_LINKS = {}

async def process_meetings(meetings, folder_path):
    for meeting in meetings:
        minutes = meeting["Minutes"]
        if isinstance(minutes, str) or not minutes.get('link'):
            continue
        print(minutes['link'])
        department = meeting["Department"]
        if not department:
            continue
        dept_name = department['value'] if isinstance(department, dict) else department
        dept_data_folder = folder_path / dept_name
        dept_data_folder.mkdir(parents=True, exist_ok=True)
        
        if isinstance(department, dict) and department.get('link') is not None:
            DEPARTMENT_LINKS[dept_name] = department['link']
        
        date_str = transform_date_format(meeting["Date"])
        time_str = meeting["Time"].split(' ')[0].replace(':', '-')
        minutes_file_path = dept_data_folder / f"{date_str}_{time_str}_minutes.pdf"
        await download_pdf_file(minutes['link'], str(minutes_file_path))

async def process_place_metadata(place_metadata_folder, place_data_folder):
    metadata_file_path = place_metadata_folder / 'download-metadata.json'
    if metadata_file_path.exists():
        with open(metadata_file_path, 'r') as f:
            metadata = json.load(f)
        print(f"Processing {metadata['place']}")

        for department, data in metadata.items():
            if department == 'place':
                continue
            if 'meetings' in data:
                await process_meetings(data['meetings'], place_data_folder)
        sleep(1000)
    else:
        print(f"metadata.json not found in {place_metadata_folder}")

async def process():
    print('creating metadata for each department')
    await process_place_metadata(place_metadata_folder, place_data_folder)

if __name__ == "__main__":
    import asyncio
    asyncio.run(process())